src/System.Memory/src/System/Buffers/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,223 @@
+        }
+
+        public int BytesWritten
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _written;
+            }
+        }
+
+        public int Capacity
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _rentedBuffer.Length;
+            }
+        }
+
+        public int BytesAvailable
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _rentedBuffer.Length - _written;
+            }
+        }
+
+        public long TotalBytesWritten
+        {
+            get
+            {
+                CheckIfDisposed();
+
+                return _totalWritten;
+            }
+        }
+
+        public void Reset()
+        {
+            CheckIfDisposed();
+
+            ResetHelper();
+        }
+
+        private void ResetHelper()
+        {
+            _rentedBuffer.AsSpan(0, _written).Clear();
+            _written = 0;
+        }
+
+        public async Task CopyToAndResetAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            CheckIfDisposed();
+
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            await stream.WriteAsync(_rentedBuffer, 0, _written, cancellationToken).ConfigureAwait(false);
+            _totalWritten += _written;
+
+            ResetHelper();
+        }
+
+        public void CopyToAndReset(Stream stream)
+        {
+            CheckIfDisposed();
+
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
+
+            stream.Write(_rentedBuffer, 0, _written);
+            _totalWritten += _written;
+
+            ResetHelper();
+        }
+
+        // Returns the rented buffer back to the pool
+        public void Dispose()
+        {
+            if (_rentedBuffer == null)
+            {
+                return;
+            }
+
+            ArrayPool<byte>.Shared.Return(_rentedBuffer, clearArray: true);
+            _rentedBuffer = null;
+            _written = 0;
+            _totalWritten = 0;
+        }
+
+        private void CheckIfDisposed()
+        {
+            if (_rentedBuffer == null)
+                throw new ObjectDisposedException(nameof(ArrayBufferWriter));
+        }
+
+        void IBufferWriter<byte>.Advance(int count)
+        {
+            CheckIfDisposed();
+
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            if (_written > _rentedBuffer.Length - count)
+                ThrowHelper.ThrowInvalidOperationException(_rentedBuffer.Length);
+
+            _written += count;
+        }
+
+        Memory<byte> IBufferWriter<byte>.GetMemory(int sizeHint)
+        {
+            CheckIfDisposed();
+
+            if (sizeHint < 0)
+                throw new ArgumentException(nameof(sizeHint));
+
+            CheckAndResizeBuffer(sizeHint);
```

Comment:
Can the above lines be moved into CheckAndResizeBuffer rather than duplicating them for both memory and span?